File: pramp/sentence_reverse/sentence_reverse.js
// Sentence Reverse
// You are given an array of characters arr that consists of sequences of characters separated by space characters. Each space-delimited sequence of characters defines a word.
//
// Implement a function reverseWords that reverses the order of the words in the array in the most efficient manner.
//
// Explain your solution and analyze its time and space complexities.
//
// Example:
//
// input:  arr = [ 'p', 'e', 'r', 'f', 'e', 'c', 't', '  ',
//                 'm', 'a', 'k', 'e', 's', '  ',
//                 'p', 'r', 'a', 'c', 't', 'i', 'c', 'e' ]
//
// output: [ 'p', 'r', 'a', 'c', 't', 'i', 'c', 'e', '  ',
//           'm', 'a', 'k', 'e', 's', '  ',
//           'p', 'e', 'r', 'f', 'e', 'c', 't' ]
// Constraints:
//
// [time limit] 5000ms
//
// [input] array.character arr
//
// 0 <= arr.length <= 100
// [output] array.character


const arr1 = ['a', 'b'];
const arr2 = ['c', 'd'];
const joinedarray = arr1.concat(arr2);

function reverseWords(arr) {

  let output = [];
  let temp = [];
  let temp_inner = [];
  arr.forEach((character, idx) => {
    if (!spaceIsHit(character)) {
      temp_inner.push(character);
      if (idx === arr.length - 1) {
        temp.push(temp_inner);
        temp_inner = [];
      }
    } else {
      temp.push(temp_inner);
      temp_inner = [];
    }
  });

  let temp_reversed = temp.reverse();

  temp_reversed.forEach((characters,idx) => {
    if (characters.length === 0) {
      output.push(' ');
      return;
    }

    output = output.concat(characters);
    if (idx !== temp_reversed.length - 1) {
      output.push(' ');
    }
  });

  return output;
}

function spaceIsHit(char) {

  if (char.trim() !== '') {
      return false;
  }

  return true;
}

File: pramp/sentence_reverse/sentence_reverse.py
#Sentence Reverse
#You are given an array of characters arr that consists of sequences of characters separated by space characters. Each space-delimited sequence of characters defines a word.
#
#Implement a function reverseWords that reverses the order of the words in the array in the most efficient manner.
#
#Explain your solution and analyze its time and space complexities.
#
#Example:
#
#input:  arr = [ 'p', 'e', 'r', 'f', 'e', 'c', 't', '  ',
#                'm', 'a', 'k', 'e', 's', '  ',
#                'p', 'r', 'a', 'c', 't', 'i', 'c', 'e' ]
#
#output: [ 'p', 'r', 'a', 'c', 't', 'i', 'c', 'e', '  ',
#          'm', 'a', 'k', 'e', 's', '  ',
#          'p', 'e', 'r', 'f', 'e', 'c', 't' ]
#Constraints:
#
#[time limit] 5000ms
#
#[input] array.character arr
#
#0 <= arr.length <= 100
#[output] array.character

def reverse_words(arr):

  # reverse the order of each word in arr
  temp_arr_inner = []
  temp_arr = []


  for idx, character in enumerate(arr):
    if idx == len(arr) - 1:
      temp_arr_inner.append(character)
      temp_arr.append(temp_arr_inner)
      break

    if space_is_hit(character):
      if len(temp_arr_inner) > 0:
        temp_arr.append(temp_arr_inner)

      temp_arr.append([' '])
      temp_arr_inner = []
    else:
      temp_arr_inner.append(character)

  reversed_temp_arr = reversed(temp_arr)


  output = sum(reversed_temp_arr, [])

  return output

def space_is_hit(character):
  if character.strip() == '':
    return True
  return False

